engine/runtime/src/main/scala/org/enso/interpreter/runtime/IrToTruffle.scala

```diff
@@ -265,101 +268,110 @@ class IrToTruffle(
-        language,
-        expressionProcessor.scope,
-        scopeBuilder.asModuleScope(),
-        expressionNode,
-        makeSection(scopeBuilder.getModule, annotation.location),
-        closureName,
-        true,
-        false
-      )
-      new RuntimeAnnotation(annotation.name, closureRootNode)
-    }
+        val argFactory =
+          new DefinitionArgumentProcessor(
+            scope       = localScope,
+            initialName = "Type " + tpDef.name.name
+          )
+        val argDefs =
+          new Array[ArgumentDefinition](atomDefn.arguments.size)
+        val argumentExpressions =
+          new ArrayBuffer[(RuntimeExpression, RuntimeExpression)]
+
+        for (idx <- atomDefn.arguments.indices) {
+          val unprocessedArg = atomDefn.arguments(idx)
+          val checkNode      = checkAsTypes(unprocessedArg)
+          val arg            = argFactory.run(unprocessedArg, idx, checkNode)
+          val fp = unprocessedArg
+            .unsafeGetMetadata(
+              FramePointerAnalysis,
+              "No frame pointer on an argument definition."
+            )
+            .asInstanceOf[FramePointer]
+          val slotIdx = fp.frameSlotIdx()
+          argDefs(idx) = arg
+          val readArgNoCheck =
+            ReadArgumentNode.build(
+              idx,
+              arg.getDefaultValue.orElse(null)
+            )
+          val readArg       = TypeCheckValueNode.wrap(readArgNoCheck, checkNode)
+          val assignmentArg = AssignmentNode.build(readArg, slotIdx)
+          val argRead =
+            ReadLocalVariableNode.build(new FramePointer(0, slotIdx))
+          argumentExpressions.append((assignmentArg, argRead))
+        }
+
+        val (assignments, reads) = argumentExpressions.unzip
+        // build annotations
+        val annotations = atomDefn.annotations.map { annotation =>
+          val scopeElements = Seq(
+            tpDef.name.name,
+            atomDefn.name.name,
+            annotation.name
+          )
+          val scopeName =
+            scopeElements.mkString(Constants.SCOPE_SEPARATOR)
+          val expressionProcessor = new ExpressionProcessor(
+            scopeName,
+            () => scopeInfo().graph,
+            () => scopeInfo().graph.rootScope,
+            dataflowInfo,
+            atomDefn.name.name,
+            frameInfo
+          )
+          val expressionNode =
+            expressionProcessor.run(annotation.expression, true)
+          val closureName = s"<default::$scopeName>"
+          val closureRootNode = ClosureRootNode.build(
+            language,
+            expressionProcessor.scope,
+            scopeBuilder.asModuleScope(),
+            expressionNode,
+            makeSection(scopeBuilder.getModule, annotation.location),
+            closureName,
+            true,
+            false
+          )
+          new RuntimeAnnotation(annotation.name, closureRootNode)
+        }
+
+        new InitializationParts(
```

Comment:
With the builder style, this would become:
```
AtomConstructor.newBuilder().
  section(...).
  scope(...).
  assignments(...).
  reads(...).
  args(defs);
```